[RDY] Improve behaviour on global changes when using system clipboard (was: vim-patch:7.4.396, 7.4.445, 7.4.598)

fmoralesc opened the pull request:
Re: issues #2793, #2584, #2155.

Work on this has to include patches [.396](https://code.google.com/p/vim/source/detail?r=7766142fc7d3), [.445](https://code.google.com/p/vim/source/detail?name=v7-4-445) and [.598](https://code.google.com/p/vim/source/detail?name=v7-4-598).
# Tasks
- [x] patch .396
- [x] patch .445
- [x] patch .598

I had to make some changes to the patches, especially .396, because of the differences in neovim's codebase. I would appreciate if someone could check if the behaviour is completely as expected.

fmoralesc committed:
clipboard: improve behavior on global changes

This is equivalent to patches 7.4.396, 7.4.445 and 7.4.598.

vim-patch:7.4.396

Problem:    When 'clipboard' is "unnamed", :g/pat/d is very slow.
(Praful)
Solution:   Only set the clipboard after the last delete. (Christian
Brabandt)

https://github.com/vim/vim/commit/1f285eb49a709e00552f7bef7e74efff5ae79026

vim-patch:7.4.445

Problem:    Clipboard may be cleared on startup.
Solution:   Set clip_did_set_selection to -1 during startup. (Christian
	    Brabandt)

https://github.com/vim/vim/commit/1a19d37d90f037c09183ba68fdddf70ab8ee179a

vim-patch:7.4.598

Problem:    ":tabdo windo echo 'hi'" causes "* register not to be
changed.
	    (Salman Halim)
Solution:   Change how clip_did_set_selection is used and add
	    clipboard_needs_update and global_change_count.  (Christian
	    Brabandt)

https://github.com/vim/vim/commit/af6a579263a688f30bfbbee72b28d08cc7e0f3d4

Co-Author: @bfredl

oni-link commented on src/nvim/ops.c:
```diff
@@ -5404,9 +5411,60 @@ static bool get_clipboard(int name, yankreg_T **target, bool quiet)
+    if (clipboard_needs_update) {
+      if (clip_unnamed & CB_UNNAMED)
+      {
+	  set_clipboard('*', &y_regs[STAR_REGISTER]);
```
During global changes `"*` and `"+` are not used (do not change). Setting the clipboard with the same content would be unnecessary. Shoudn't we use register `""` to update register `"*` and/or `"+`?

bfredl replied:
Correct. I think we could replace both inner if's here with a single `set_clipboard(NUL, y_current);` (which should handle all cases, including `cb_flags == 0`, and no change will be needed here when #3096 gets fixed, I will look into that one tomorrow).

splinterofchaos commented on src/nvim/ops.c:
```diff
@@ -72,6 +72,10 @@ static yankreg_T y_regs[NUM_REGISTERS];
+
+static int clipboard_needs_update; // clipboard needs to be updated
+static int global_change_count = 0; // if set, inside global changes
+				    //(start_global_changes() ... end_global_changes())
```
Looks like the alignment is one off.

justinmk replied:
should be spaces not tabs :)

bfredl commented on src/nvim/option.c:
```diff
@@ -2578,10 +2578,32 @@ did_set_string_option (
       errmsg = e_invarg;
-  } else if (varp == &(curwin->w_s->b_p_spl)
+    }
+    if (errmsg == NULL) {
```
Isn't this redundant with the existing `opt_string_flags` ?

fmoralesc replied:
I didn't see `clip_unnamed` was replaced with `cb_flags`, so I tried to restore it here. I'll remove this.

bfredl commented on src/nvim/ops.c:
```diff
@@ -5404,9 +5411,60 @@ static bool get_clipboard(int name, yankreg_T **target, bool quiet)
+  clipboard_needs_update = FALSE;
+
+  if (clip_did_set_selection) {
+    clip_unnamed = 0;
```
`clip_unnamed` did become `cb_flags & CB_UNNAMEDMASK`  in the clipboard refactor. But rather than saving, hiding, and then restoring the value, can't `adjust_clipboard_name` (the only place where `cb_flags` have direct effect) just check the relevant condition directly ?

fmoralesc commented:
Rebased and fixed the biggest issues this had. In the end, this became very different to the original vim patches, so I squashed int a single commit (should be simpler to review). The variables that track the state are no longer global, and I think `clip_did_set_selection` is not really needed anymore. I'll look into @bfredl's suggestion to turn [this conditional](https://github.com/neovim/neovim/pull/2809/files#diff-8b8e0501c81be28ef576fb23580d1125R5450) into this:
~~~ c
    set_clipboard(NUL, y_previous);
~~~

bfredl commented:
`clip_did_set_selection` isn't really used yes, but I think it needs to be. The new checks are a bit overzealous: With `clipboard=unnamed`, just `:g/re/p` (print) might overwrite primary selection. also `:g/re/put` is broken, as is `:g/re/y *` _without_ unnamed .

fmoralesc commented:
What do you suggest, @bfredl? I won't have time to take a look at this for a while.

bfredl commented:
I did a bit of [experimentation](https://github.com/bfredl/neovim/commit/1f360a5928e70a4aad0fbee3a3220b6b72c59f13) of more fine-grained checks. It handles the cases above at least, but there might be others. A few `:g/.../...` tests added to the clipboard_provider_spec might be good. (For instance, to test the main point, we could add a counter to `s:methods.set`  in `provider#clipboard`  and check that `g/.../d` doesn't cause too many invokes )

fmoralesc commented:
@bfredl I just tried your changes, and it seems they regress on this on the `g/re/y *` scenario, since the clipboard is accessed on every iteration.

bfredl commented:
Looking at the original patches, I don't think they did speedup this either, (`g/re/y *` is quite contrived after all, or at least more obvious it can be expensive, while `g/re/d` was the actual problem users reported). Nor did this PR either, that is without expense of correctness (`cb=unnamed` and `g/re/y +` yanks to primary and not clipboard).
To fully support `y */+` speedup as well, you would need individual `did_set_` flags for each register, but I don't really think that is needed (all reported issues were about unnamed delete after all)

fmoralesc commented:
@bfredl I would be happy to include your changes to this PR and add some tests. As I said, I don't think I'll be able to look much more into this, so the quickest resolution seems like the best to me.

I was thinking `end_global_changes` could check for changes on `y_previous` before trying to set the clipboard, that could make `clip_did_set_selection` unneeded.

bfredl commented on src/nvim/ops.c:
```diff
@@ -72,6 +72,12 @@ static yankreg_T y_regs[NUM_REGISTERS];
 
 static bool clipboard_didwarn_unnamed = false;
+
+static bool clipboard_needs_update = true;  // clipboard needs to be updated
```
I think `clipboard_needs_update` is quite confusing, to me it sounds like it does what `clip_did_set_selection` does (the register was set internally and thus clipboard needs update). Perhaps negating it and `clipboard_delay_set` or `_copy` or something, to name it after the effect it has.

bfredl commented:
I can for the love of not understand the purpose of `clip_did_set_selection` in the original patch series*, but it seems like it uses `clipboard_needs_update` as I expected in the comment above :) So in summary, on top of my exisiting changes I would suggest rename
`clipboard_needs_update` => `clipboard_delay_set`
`clip_did_set_selection` (as used in my changes) => `clipboard_needs_update`

( * nor the purpose of code like this
```
    if (*rp == 0 && (clip_unnamed != 0 || clip_unnamed_saved != 0))
    {
	if (clip_unnamed != 0)
	    *rp = ((clip_unnamed & CLIP_UNNAMED_PLUS) && clip_plus.available)
								  ? '+' : '*';
	else
	    *rp = ((clip_unnamed_saved & CLIP_UNNAMED_PLUS) && clip_plus.available)
								  ? '+' : '*';
    }
```
but that's an unrelated disscussion )

fmoralesc commented:
Sure. I renamed `clipboard_needs_update` to `delay_clipboard_update`, it sounded better to me.

I'll squash the whole set of changes once I'm done and mark you as co-author. But first, I'll try to add some tests.

justinmk commented on src/nvim/ops.c:
```diff
@@ -72,6 +72,12 @@ static yankreg_T y_regs[NUM_REGISTERS];
+// for behavior between start_global_changes() and end_global_changes())
+static bool delay_clipboard_update = false;  // clipboard's update needs to be delayed
+static int global_change_count = 0;          // if set, inside global changes
+static bool clipboard_needs_update = true;   // the clipboard was updated
```
For our sanity it would help to namespace flags like this. We have `clipboard_didwarn_unnamed` and `clipboard_needs_update`, so `clipboard_delay_update` would remove some entropy.

justinmk replied:
In general in our C code, the pattern is "noun-verb-foo".

fmoralesc replied:
:+1:

bfredl commented on src/nvim/ops.c:
```diff
@@ -5404,9 +5418,44 @@ static bool get_clipboard(int name, yankreg_T **target, bool quiet)
 }
 
+
+/*
```
could merge this two comment blocks and into doxygen style. Also I would suggest moving `start_`/`end_` below `set_clipboard` ( as `adjust_clipboard_name`, `get_clipboard`, and `set_clipboard` belong close together, while `*_global_changes` is more of a special case)

fmoralesc commented:
Just noticed... @bfredl, your commit broke the clipboard_spec test. :/ 

(BTW, I'm having an issue with the generation of the tags file in my local builds, it's preventing me to run the tests normally. cc @justinmk)

bfredl commented:
I think because I forgot to update initilalization, with new names `clipboard_needs_update` should initialize to `false`

fmoralesc commented:
Yeah, that solves it.

bfredl commented on test/functional/clipboard/clipboard_provider_spec.lua:
```diff
@@ -285,6 +285,22 @@ describe('clipboard usage', function()
 
+    it('is updated on global changes', function()
+      insert([[
+      text
```
indentation

bfredl commented on src/nvim/testdir/test_eval.in:
```diff
@@ -122,7 +121,19 @@ call SetReg('/', ['abc/'])
+:let &cb='unnamedplus'
+:1y
+:AR +
+:tabdo :windo :echo "hi"
```
Is this addition mainly to test the `:tabdo :windo` bug? Then maybe it could be converted to a new-style test that doesn't clobber system clipboard. (If not please disregard this comment)

fmoralesc replied:
Now that I think about it, this test does nothing for us (it always succeeds, since the clipboard provider is never loaded for this test). Also, now that I *really* think about it, why *would* `:tabdo windo echo 'hi'` touch the `*` register? Makes no sense to me.

bfredl replied:
I think it was a bug in vim that `:tabdo :windo` would call `begin_`/`end_global_changes` recursively before that was supported, fixed by 598 . Do you mean that rtp is setup in the oldtests so that `autoload/provider/clipboard.vim` is not in `rtp` for the tests ? That would be fixed anyway by converting to lua test and use the shim provider.

fmoralesc replied:
The oldtests use `-u NONE`, so nothing in the runtime is loaded, including the providers.

I think the bulk of the test is more or less redundant with the tests already in place in the clipboard spec test.

bfredl replied:
Really? When I do `nvim -u NONE` I can still use the clipboard. True, Considering that we have changed the clipboard implementation so much anyway, this test for a very specific bug in earlier version of the vim implementation might arguably not be so relevant, but if we do want keep it, more or less repeat lines 296-299
```
      insert("one two")
      feed('^"+dwdw"+P')
      expect('two')
      eq({{'two'}, 'v'}, eval("g:test_clip['+']"))
```
but with a `execute(':tabdo :windo :echo "hi" ')` in front.

fmoralesc replied:
Ah, sorry, I forgot the providers are autoloaded, so they still get pulled.

fmoralesc replied:
I was checking what was missing for this, and I thought this test wasn't particularly meaningful, so I'm leaving things as-is.

bfredl commented on test/functional/clipboard/clipboard_provider_spec.lua:
```diff
@@ -329,6 +345,21 @@ describe('clipboard usage', function()
     end)
+    it('is updated on global changes', function()
+      insert([[
+      text
```
indentation

fmoralesc commented:
Rebased to master, fixed latest points. Marking as RDY.

justinmk commented:
Merged
